File: cuecard-mobile/ios/CueCard/CueCard/Modifiers/GlassEffect.swift
import SwiftUI

// MARK: - Glass Effect Modifier for iOS 26+

extension View {
    /// Applies a glass effect to the view.
    /// - Parameters:
    ///   - shape: The shape to apply the glass effect in
    ///   - interactive: Whether the glass should respond to user interactions (iOS 26+ only)
    /// - Returns: A view with the glass effect applied
    @ViewBuilder
    func glassedEffect<S: Shape>(in shape: S = Capsule(), interactive: Bool = false) -> some View {
        if #available(iOS 26.0, *) {
            self.glassEffect(interactive ? .regular.interactive() : .regular, in: shape)
        } else {
            self.background {
                shape.glassed()
            }
        }
    }
}

// MARK: - Fallback Glass Effect for older iOS versions

extension Shape {
    /// Creates a glass-like appearance for older iOS versions using materials and gradients
    func glassed() -> some View {
        self
            .fill(.ultraThinMaterial)
            .overlay {
                self.fill(
                    .linearGradient(
                        colors: [
                            .primary.opacity(0.08),
                            .primary.opacity(0.05),
                            .primary.opacity(0.01),
                            .clear,
                            .clear,
                            .clear
                        ],
                        startPoint: .topLeading,
                        endPoint: .bottomTrailing
                    )
                )
            }
            .overlay {
                self.stroke(.primary.opacity(0.2), lineWidth: 0.7)
            }
    }
}

// MARK: - Script Edge Fade

extension View {
    /// Fades a scrolling script into the background at its top and bottom edges,
    /// so lines arrive and leave instead of being cut off mid-stroke.
    ///
    /// The fades are painted in the background color rather than masked, so they
    /// cover the text without touching whatever floats above them.
    func scriptEdgeFade(for colorScheme: ColorScheme, top: CGFloat, bottom: CGFloat) -> some View {
        let color = AppColors.background(for: colorScheme)

        return overlay(alignment: .top) {
            EdgeFade(color: color, height: top, isTop: true)
        }
        .overlay(alignment: .bottom) {
            EdgeFade(color: color, height: bottom, isTop: false)
        }
    }
}

/// One end of the fade: solid against its edge, gone by the far side.
private struct EdgeFade: View {
    let color: Color
    let height: CGFloat
    let isTop: Bool

    var body: some View {
        LinearGradient(
            stops: [
                .init(color: color, location: 0),
                .init(color: color.opacity(0.9), location: 0.45),
                .init(color: color.opacity(0), location: 1)
            ],
            startPoint: isTop ? .top : .bottom,
            endPoint: isTop ? .bottom : .top
        )
        .frame(height: height)
        .allowsHitTesting(false)
    }
}

File: cuecard-mobile/android/app/src/main/java/com/thisisnsh/cuecard/android/Modifiers/GlassEffect.kt
package com.thisisnsh.cuecard.android.modifiers

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.compositeOver
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.thisisnsh.cuecard.android.models.AppColors

/** The capsule shape SwiftUI's `Capsule()` draws. */
val Capsule: Shape = RoundedCornerShape(percent = 50)

/**
 * The look the app's floating controls sit on, ported from the SwiftUI
 * `Shape.glassed()`.
 *
 * Compose has no backdrop blur — `Modifier.blur` blurs a composable's own content,
 * not what sits behind it — so there is nothing here for a scrim of the page color
 * to frost. Painting one only tinted the page with itself and left the controls
 * reading as flat outlines. Instead the surface is the page lifted toward the text
 * color: an opaque fill that stands off the page the way the material does on iOS.
 *
 * `tint` gives a control its own fill — the green play buttons — and then nothing
 * is painted over it, so the color stays the color it was asked for. The gradient
 * and the stroke are the SwiftUI ones either way.
 */
fun Modifier.glassed(
    shape: Shape = Capsule,
    isDark: Boolean,
    tint: Color? = null
): Modifier {
    val primary = AppColors.textPrimary(isDark)
    val surface = tint ?: primary
        .copy(alpha = if (isDark) 0.16f else 0.10f)
        .compositeOver(AppColors.background(isDark))

    return this
        .clip(shape)
        .background(surface, shape)
        .background(
            brush = Brush.linearGradient(
                colors = listOf(
                    primary.copy(alpha = 0.08f),
                    primary.copy(alpha = 0.05f),
                    primary.copy(alpha = 0.01f),
                    Color.Transparent,
                    Color.Transparent,
                    Color.Transparent
                ),
                start = Offset.Zero,
                end = Offset.Infinite
            ),
            shape = shape
        )
        .border(width = 0.7.dp, color = primary.copy(alpha = 0.2f), shape = shape)
}

/**
 * Fades a scrolling script into the background at its top and bottom edges, so
 * lines arrive and leave instead of being cut off mid-stroke.
 *
 * The fades are painted in the background color rather than masked, so they cover
 * the text without touching whatever floats above them.
 */
fun Modifier.scriptEdgeFade(isDark: Boolean, top: Dp, bottom: Dp): Modifier =
    drawWithContent {
        drawContent()

        val color = AppColors.background(isDark)
        val topHeight = top.toPx()
        val bottomHeight = bottom.toPx()

        if (topHeight > 0f) {
            drawRect(
                brush = Brush.verticalGradient(
                    0f to color,
                    0.45f to color.copy(alpha = 0.9f),
                    1f to color.copy(alpha = 0f),
                    startY = 0f,
                    endY = topHeight
                ),
                size = Size(size.width, topHeight)
            )
        }

        if (bottomHeight > 0f) {
            drawRect(
                brush = Brush.verticalGradient(
                    0f to color.copy(alpha = 0f),
                    0.55f to color.copy(alpha = 0.9f),
                    1f to color,
                    startY = size.height - bottomHeight,
                    endY = size.height
                ),
                topLeft = Offset(0f, size.height - bottomHeight),
                size = Size(size.width, bottomHeight)
            )
        }
    }
